File: poisson_solver.f90
program poisson_solver
    implicit none
    integer, parameter :: n = 500
    integer, parameter :: max_iter = 500
    real(8), dimension(n, n) :: u, u_new, f
    real(8) :: dx, x, y, start_time, end_time
    integer :: i, j, iter

    ! Initialize grids
    u = 0.0d0
    u_new = 0.0d0
    f = 0.0d0

    ! Set up source term
    dx = 1.0d0 / (n - 1)
    do i = 1, n
        do j = 1, n
            x = (i - 1) * dx
            y = (j - 1) * dx
            f(i, j) = sin(x * 3.14159265358979323846d0) * sin(y * 3.14159265358979323846d0)
        end do
    end do

    print '(A,I0,A,I0,A,I0,A)', 'Fortran Poisson Solver: ', n, 'x', n, ' grid, ', max_iter, ' iterations'

    call cpu_time(start_time)

    ! Jacobi iteration
    do iter = 1, max_iter
        do i = 2, n-1
            do j = 2, n-1
                u_new(i, j) = 0.25d0 * (u(i+1, j) + u(i-1, j) + &
                                        u(i, j+1) + u(i, j-1) + &
                                        dx * dx * f(i, j))
            end do
        end do

        ! Swap arrays
        u = u_new
    end do

    call cpu_time(end_time)

    print '(A,F0.4,A)', 'Time: ', end_time - start_time, ' seconds'
    print '(A,F0.6)', 'Sample value at center: ', u(n/2, n/2)

end program poisson_solver

File: poisson_solver.py
#!/usr/bin/env python3
import time
import numpy as np

def jacobi_poisson_solver(n, max_iter):
    """
    Solve 2D Poisson equation using Jacobi iteration
    Boundary conditions: u = 0 on edges
    Source term: f = sin(x) * sin(y)
    """
    # Initialize grid
    u = np.zeros((n, n))
    u_new = np.zeros((n, n))
    f = np.zeros((n, n))

    # Set up source term
    dx = 1.0 / (n - 1)
    for i in range(n):
        for j in range(n):
            x = i * dx
            y = j * dx
            f[i, j] = np.sin(x * np.pi) * np.sin(y * np.pi)

    # Jacobi iteration
    for iteration in range(max_iter):
        for i in range(1, n-1):
            for j in range(1, n-1):
                u_new[i, j] = 0.25 * (u[i+1, j] + u[i-1, j] +
                                       u[i, j+1] + u[i, j-1] +
                                       dx * dx * f[i, j])

        # Swap arrays
        u, u_new = u_new, u

    return u

if __name__ == "__main__":
    n = 500
    max_iter = 500

    print(f"Python Poisson Solver: {n}x{n} grid, {max_iter} iterations")

    start = time.time()
    u = jacobi_poisson_solver(n, max_iter)
    end = time.time()

    print(f"Time: {end - start:.4f} seconds")
    print(f"Sample value at center: {u[n//2, n//2]:.6f}")
